fe/fe-core/src/main/java/org/apache/doris/nereids/pattern/GroupExpressionMatching.java

```diff
@@ -0,0 +1,130 @@
+// Licensed to the Apache Software Foundation (ASF) under one
+// or more contributor license agreements.  See the NOTICE file
+// distributed with this work for additional information
+// regarding copyright ownership.  The ASF licenses this file
+// to you under the Apache License, Version 2.0 (the
+// "License"); you may not use this file except in compliance
+// with the License.  You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+package org.apache.doris.nereids.pattern;
+
+import org.apache.doris.nereids.memo.GroupExpression;
+import org.apache.doris.nereids.trees.TreeNode;
+import org.apache.doris.nereids.trees.Utils;
+
+import com.google.common.collect.Lists;
+
+import java.util.Iterator;
+import java.util.List;
+import java.util.NoSuchElementException;
+import java.util.Objects;
+
+/**
+ * Get all pattern matching subtree in query plan from a group expression.
+ */
+public class GroupExpressionMatching<NODE_TYPE extends TreeNode> implements Iterable<NODE_TYPE> {
+    private final Pattern pattern;
+    private final GroupExpression groupExpression;
+
+    public GroupExpressionMatching(Pattern pattern, GroupExpression groupExpression) {
+        this.pattern = Objects.requireNonNull(pattern);
+        this.groupExpression = Objects.requireNonNull(groupExpression);
+    }
+
+    @Override
+    public GroupExpressionIterator<NODE_TYPE> iterator() {
+        return new GroupExpressionIterator<>(pattern, groupExpression);
+    }
+
+    /**
+     * Iterator to get all subtrees.
+     */
+    public static class GroupExpressionIterator<NODE_TYPE extends TreeNode> implements Iterator<NODE_TYPE> {
+        private final List<NODE_TYPE> results;
```

Comment:
Save all combination plan in the iterator maybe oom, because the exponential relationship between the number of plan combinations and the number of layers. We should optimize it later.

Reply:
yes, will change it in later PR